server/src/main/java/org/candlepin/dto/api/v1/ActivationKeyDTO.java

```diff
@@ -0,0 +1,954 @@
+    }
+
+    /**
+     * Adds the given pool to this activation key DTO.
+     *
+     * @param pool the pool to add to this activation key DTO.
+     *
+     * @return true if this pool was not already contained in this activation key DTO.
+     */
+    @XmlTransient
+    public boolean addPool(ActivationKeyPool pool) {
+        if (pool == null || pool.getPool() == null ||
+            pool.getPool().getId() == null || pool.getPool().getId().isEmpty()) {
+            throw new IllegalArgumentException("pool is null or has no id.");
+        }
+
+        if (this.pools == null) {
+            this.pools = new HashSet<ActivationKeyPoolDTO>();
+        }
+
+        ActivationKeyPoolDTO akp = new ActivationKeyPoolDTO(pool.getPool().getId(), pool.getQuantity());
+
+        return this.pools.add(akp);
+    }
+
+    /**
+     * Retrieves a view of the products for the activation key represented by this DTO. If the products
+     * have not yet been defined, this method returns null.
+     * <p></p>
+     * Note that the collection returned by this method is a view of the collection backing this
+     * set of products. Elements cannot be added to the collection, but elements may be removed.
+     * Changes made to the collection will be reflected by this product data instance.
+     *
+     * @return
+     *  the products associated with this key, or null if the products have not yet been defined
+     */
+    @JsonIgnore
+    public Set<String> getProducts() {
+        return this.products != null ? new SetView<String>(this.products) : null;
+    }
+
+    /**
+     * Adds the collection of products to this ActivationKey DTO.
+     * @param products
+     *  A set of products to attach to this DTO, or null to clear the existing ones
+     * @return
+     *  A reference to this DTO
+     */
+    @JsonIgnore
+    public ActivationKeyDTO setProducts(Set<String> products) {
+        if (products != null) {
+            if (this.products == null) {
+                this.products = new HashSet<String>();
+            }
+            else {
+                this.products.clear();
+            }
+
+            for (String dto : products) {
+                if (dto == null || dto.isEmpty()) {
+                    throw new IllegalArgumentException("collection contains null or incomplete dtos");
+                }
+            }
+
+            this.products.addAll(products);
+        }
+        else {
+            this.products = null;
+        }
+        return this;
+    }
+
+    /**
+     * Strictly used only for serialization to maintain the output format:
+     * <pre>
+     * {@code   "products": [
+     *          {
+     *              "productId": "string"
+     *          }
+     *      ]
+     *}</pre>
+     *
+     * @return
+     *  a set of maps that each has a single entry, whose key is 'productId'
+     *  and value is the value of the product id.
+     */
+    @JsonProperty("products")
+    public Set<Map<String, String>> getProductDTOs() {
```

Comment:
This is a good start, but it starts muddying our API. An alternative would be to write a serializer and deserializer and apply those to the standard methods. An existing example of this is currently occurring in Pool with attributes, which uses org.candlepin.jackson.CandlepinLegacyAttributeSerializer and .CandlepinAttributeDeserializer. This way we can get away from duplicating the entire data structure and can, instead, write it out directly.